lib/active_model/serializer/reflection.rb

```diff
@@ -78,9 +86,15 @@ def value(serializer)
           if block_value != :nil
```

Comment:
@NullVoxPopuli I know @bf4 knows this area better than I do, but wanted to address what I think one key problematic area is, and how it relates to https://github.com/rails-api/active_model_serializers/pull/1720:

```ruby
# evaluates the relationship
# - this is only needed if we want to use 'includes', otherwise, we only need ids.
association_value = value(subject)
reflection_options = options.dup

# we only need a serializer if we are using includes
# without includes, all we need are the id and type
```

The issue I see here is sometimes you want a `link`, not `id/type` combo. In this case:

* We need to fire `#value` to A) see if we want `id/type` in the response (`include_data`) and B) get the actual links data.
* This will run the association block
* If you have an in-line association, you're now firing the unnecessary SQL again.

In code:

```ruby
belongs_to :author do
  link :related do
    href "/foo/bar"
  end

  # we only want to run this code if :author is included, but instead it
  # has to run in any case to get the link href above
  Author.first || Author.default
end
```

> I guess what we could do is maybe have support for looking up custom types / polymorphic types via ActiveRecord if that is available, and then also provide a way to override the serialized type name per relationship in the serializer?

Unfortunately I think this would cause the same inline-association issue. Just for the sake of argument, let's say the solution was a `type` line, you'd have something like this:

```ruby
belongs_to :author do
  type 'users'

  Author.first || Author.default
end
```

In order to get the `type` value, you need to run the block, which is going to fire the unnecessary SQL again.

This is the why https://github.com/rails-api/active_model_serializers/pull/1720 changes inline-associations to be lazily evaluated by wrapping the data loading in a proc:

```ruby
belongs_to :author do
  type 'users'
  
  # we can now conditionally/lazily fire this code
  load_data { Author.first || Author.default }
end
```